### PR TITLE
Updated BLAS1 tests to match BLAS2/3 tests

### PR DESCRIPTION
**Pull request by RossBrunton**
This uses a parametized test now, and compares their values to
those produced by the reference blas (rather than hardcoding the
algorithm).

### CODE REVIEW

**Comment by RossBrunton** [comment]
Locally the rotg test seems to be failing. Lets see what Travis says about that.

**Comment by mehdi-goli** [comment]
Require to fix the precision issue for Intel:GPU

File: test/blas_test.hpp [hunk 1]
```diff
@@ -204,8 +204,13 @@ class BLAS_Test<blas_test_args<ScalarT_, MetadataT_, ExecutorType_>>
   }
 };
 
-#define ASSERT_T_EQUAL(T, val1, val2)                                         \
-  ASSERT_PRED_FORMAT2(::testing::internal::CmpHelperFloatingPointEQ<T>, val1, \
-                      val2)
+constexpr const int ulp_required_in_assert = 20;
+
+#define ASSERT_T_EQUAL(T, val1, val2)                                   \
```
Comment (mehdi-goli) [inline test/blas_test.hpp]: create a function called precision Computation and return the precision. So in this case you dont need macro for asser_near and you can directly put it in the code

File: test/blas_test.hpp [hunk 2]
```diff
@@ -204,8 +204,13 @@ class BLAS_Test<blas_test_args<ScalarT_, MetadataT_, ExecutorType_>>
...
+  do {                                                                  \
+    T precision = std::numeric_limits<T>::epsilon() *                   \
+                  std::abs((val1) + (val2)) * ::ulp_required_in_assert; \
```
Comment (Nyrio) [inline test/blas_test.hpp]: Shouldn't it be `fabs`?

Comment (Nyrio) [reply test/blas_test.hpp]: Also if you can put a link or explanation of where that 20 comes from that would be nice

Comment (RossBrunton) [reply test/blas_test.hpp]: std::abs and std::fabs do the same thing if cmath is imported. Because C++.

File: test/blas_test.hpp (hunk 1 above)
Comment (RossBrunton) [reply test/blas_test.hpp]: Should be done now.

File: test/blas_test.hpp (hunk 2 above)
Comment (Nyrio) [reply test/blas_test.hpp]: Are you absolutely sure about this? I don't see floating point versions there: https://en.cppreference.com/w/cpp/numeric/math/abs

**Comment by mehdi-goli** [comment]
Can you please the std::uniform distribution for data as used in the benchmark. Also, can you use the new floating point precision validation algorithm for test

**Comment by Nyrio** [comment]
@RossBrunton this is the code used for the benchmarks:
```cpp
#include <random>
...

/**
 * @fn random_data
 * @brief Generates a random vector of scalar values, using a uniform
 * distribution.
 */
template <typename scalar_t>
static inline std::vector<scalar_t> random_data(size_t size,
                                                bool initialized = true) {
  std::vector<scalar_t> v = std::vector<scalar_t>(size);

  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(-2.0, 2.0);
  for (int i = 0; i < v.size(); ++i) {
    v[i] = dis(gen);
  }
  return v;
}
```

File: test/blas_test.hpp [hunk 3]
```diff
@@ -204,8 +204,14 @@ class BLAS_Test<blas_test_args<ScalarT_, MetadataT_, ExecutorType_>>
   }
 };
 
-#define ASSERT_T_EQUAL(T, val1, val2)                                         \
-  ASSERT_PRED_FORMAT2(::testing::internal::CmpHelperFloatingPointEQ<T>, val1, \
-                      val2)
+// The "distance" required between two floats for precision_limit
+// Higher numbers = less precision, 40 was obtained through experimentation
+// This results in values in xe-5 range for floats, and xe-14 for doubles
+constexpr const int ulp_required_for_precision = 40;
```
[the remark this answers is not in the document]
Comment (RossBrunton) [reply test/blas_test.hpp]: This is no longer relevant.

File: test/unittest/blas3/blas3_gemm_test.cpp [hunk 4]
```diff
@@ -116,15 +116,14 @@ void run_test(const combination_t<scalar_t> combi) {
...
+        policy_handler.copy_to_host(m_c_gpu + _base(dim_c, ldc_mul, bs),
+                                    c_m_gpu.data() + _base(dim_c, ldc_mul, bs),
+                                    _size_batch(dim_c, ldc_mul));
     policy_handler.wait(event);
   }
 
-  for (int i = 0; i < _size(dim_c, ldc_mul); ++i) {
-    ASSERT_T_EQUAL(scalar_t, c_m_gpu[i], c_m_cpu[i]);
-  }
+  ASSERT_TRUE(utils::compare_vectors(c_m_gpu, c_m_cpu));
```
Comment (mehdi-goli) [inline test/unittest/blas3/blas3_gemm_test.cpp]: You can call this ASSERT_NEAR or ASSERT_ALMOST_EQUAL

Comment (RossBrunton) [reply test/unittest/blas3/blas3_gemm_test.cpp]: What do you mean? Do you want me to get the threshold and use ASSERT_NEAR on it?

Comment (Nyrio) [reply test/unittest/blas3/blas3_gemm_test.cpp]: We discussed it offline, you can ignore that

**Pull request merged**